### ISSUE TITLE
Crash report - width and height must be non-negative

### ISSUE THREAD

**Issue by JuliusSweetland** [JuliusSweetland/OptiKey#79]
````
INFO  JuliusSweetland.OptiKey.App: STARTING UP.
2015-09-10 13:42:24,688 [ 1] INFO  JuliusSweetland.OptiKey.App: Assembly version: 2.0.5.0
2015-09-10 13:42:24,692 [ 1] INFO  JuliusSweetland.OptiKey.App: Process elevated: False
2015-09-10 13:42:24,692 [ 1] INFO  JuliusSweetland.OptiKey.App: Process bitness: 32-Bit
2015-09-10 13:42:24,693 [ 1] INFO  JuliusSweetland.OptiKey.App: OS version: Windows 10
2015-09-10 13:42:24,693 [ 1] INFO  JuliusSweetland.OptiKey.App: OS service pack: No Service Pack
2015-09-10 13:42:24,693 [ 1] INFO  JuliusSweetland.OptiKey.App: OS bitness: 64-Bit
2015-09-10 13:42:24,934 [ 1] INFO  JuliusSweetland.OptiKey.App: Boot strapping the services and UI.
2015-09-10 13:42:25,377 [ 1] INFO  JuliusSweetland.OptiKey.Services.PublishService: Releasing all keys (with virtual key codes) which are down.
2015-09-10 13:42:25,498 [ 1] INFO  JuliusSweetland.OptiKey.Services.PublishService: Releasing all keys (with virtual key codes) which are down.
2015-09-10 13:42:25,886 [ 1] INFO  JuliusSweetland.OptiKey.Services.InputService: PointsPerSecond event has first subscriber.
2015-09-10 13:42:25,948 [ 1] INFO  JuliusSweetland.OptiKey.Services.InputService: CurrentPosition event has first subscriber.
2015-09-10 13:42:25,957 [ 1] INFO  JuliusSweetland.OptiKey.Services.InputService: SelectionProgress event has first subscriber.
2015-09-10 13:42:25,987 [ 1] INFO  JuliusSweetland.OptiKey.Services.InputService: Selection event has first subscriber.
2015-09-10 13:42:25,988 [ 1] INFO  JuliusSweetland.OptiKey.Services.InputService: Selection Result event has first subscriber.
2015-09-10 13:42:30,328 [ 1] ERROR JuliusSweetland.OptiKey.App: A DispatcherUnhandledException has been encountered...
System.ArgumentException: Width and Height must be non-negative.
   at System.Windows.Rect..ctor(Double x, Double y, Double width, Double height)
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.SetAppBarSizeAndPosition(DockEdges dockPosition, Rect sizeAndPosition) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 1104
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.AppBarPositionChangeCallback(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 509
   at System.Windows.Interop.HwndSource.PublicHooksFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
2015-09-10 13:42:30,484 [ 1] ERROR JuliusSweetland.OptiKey.App: A DispatcherUnhandledException has been encountered...
System.ArgumentException: Width and Height must be non-negative.
   at System.Windows.Rect..ctor(Double x, Double y, Double width, Double height)
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.SetAppBarSizeAndPosition(DockEdges dockPosition, Rect sizeAndPosition) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 1104
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.AppBarPositionChangeCallback(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 509
   at System.Windows.Interop.HwndSource.PublicHooksFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean&amp; handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
````

**Comment by JuliusSweetland** [comment]
Another example:

<?xml version="1.0"?>
<SerializableException xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <ExtendedInformation>
    <HResult>-2147024809</HResult>
  </ExtendedInformation>
  <Message>Width and Height must be non-negative.</Message>
  <Source>WindowsBase</Source>
  <StackTrace>   at System.Windows.Rect..ctor(Double x, Double y, Double width, Double height)
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.CalculateDockSizeAndPositionInPx(DockEdges position, DockSizes size) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 599
   at JuliusSweetland.OptiKey.Services.WindowManipulationService.Shrink(ShrinkFromDirections direction, Double amountInPx) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\WindowManipulationService.cs:line 490
   at JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel.HandleFunctionKeySelectionResult(KeyValue singleKeyValue) in c:\Users\Julius\Documents\GitHub\OptiKey\src\UI\ViewModels\MainViewModel.ServiceEventHandlers.cs:line 1091
   at JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel.KeySelectionResult(Nullable`1 singleKeyValue, List`1 multiKeySelection) in c:\Users\Julius\Documents\GitHub\OptiKey\src\UI\ViewModels\MainViewModel.ServiceEventHandlers.cs:line 139
   at JuliusSweetland.OptiKey.UI.ViewModels.MainViewModel.&lt;AttachServiceEventHandlers&gt;b__6(Object o, Tuple`4 tuple) in c:\Users\Julius\Documents\GitHub\OptiKey\src\UI\ViewModels\MainViewModel.ServiceEventHandlers.cs:line 112
   at JuliusSweetland.OptiKey.Services.InputService.PublishSelectionResult(Tuple`4 selectionResult) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\InputService.cs:line 404
   at JuliusSweetland.OptiKey.Services.InputService.ProcessSelectionTrigger(TriggerSignal triggerSignal) in c:\Users\Julius\Documents\GitHub\OptiKey\src\Services\InputService.subscriptions.cs:line 185
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value)
   at System.Reactive.Concurrency.ObserveOn`1.ObserveOnSink.OnNextPosted(Object value)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)</StackTrace>
  <TargetSite>Void .ctor(Double, Double, Double, Double) @ System.Windows.Rect</TargetSite>
  <Type>System.ArgumentException</Type>
</SerializableException>

**Comment by DanielHutt** [comment]
Are there any further steps to reproduce this?

**Comment by JuliusSweetland** [comment]
Just the exception details I'm afraid. I think I saw one report where the action was shrink, suggesting that someone had shrunk OptiKey in one direction so that it tried to go through 0, i.e. thinner than zero thickness. This looks like something else though, although it's possible that they did that and it saved their new (negative) width/height and then they started OptiKey after it crashed and it attempted to reinstate the meaningless dimensions. I'm not sure to be honest.

**Comment by JuliusSweetland** [comment]
Potentially fixed in v2.0.8 (at least a lot more useful logging around this area has been added so if it recurs I will have a better idea)
